adapters/missena/missena.go

```diff
@@ -0,0 +1,226 @@
+	GDPRConsent      string `json:"consent_string"`
+	GDPR             bool   `json:"consent_required"`
+	Placement        string `json:"placement"`
+	TestMode         string `json:"test"`
+}
+
+type MissenaBidServerResponse struct {
+	Ad        string  `json:"ad"`
+	Cpm       float64 `json:"cpm"`
+	Currency  string  `json:"currency"`
+	RequestId string  `json:"requestId"`
+}
+
+type MissenaInternalParams struct {
+	ApiKey           string
+	RequestId        string
+	Timeout          int
+	Referer          string
+	RefererCanonical string
+	GDPRConsent      string
+	GDPR             bool
+	Placement        string
+	TestMode         string
+}
+
+type MissenaAdapter struct {
+	EndpointTemplate *template.Template
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) makeParameter(missenaParams MissenaInternalParams, request *openrtb2.BidRequest) *MissenaAdRequest {
+	missenaRequest := MissenaAdRequest{
+		RequestId:        request.ID,
+		Timeout:          2000,
+		Referer:          request.Site.Page,
+		RefererCanonical: request.Site.Domain,
+		GDPRConsent:      missenaParams.GDPRConsent,
+		GDPR:             missenaParams.GDPR,
+		Placement:        missenaParams.Placement,
+		TestMode:         missenaParams.TestMode,
+	}
+	return &missenaRequest
+}
+
+func (a *adapter) makeRequest(missenaParams MissenaInternalParams, reqInfo *adapters.ExtraRequestInfo, existingRequests []*adapters.RequestData, request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+
+	url := a.endpoint + "?t=" + missenaParams.ApiKey
+	parameter := a.makeParameter(missenaParams, request)
+	body, errm := json.Marshal(parameter)
+	if errm != nil {
+		return nil, errm
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		headers.Add("User-Agent", request.Device.UA)
+	}
+
+	if request.Device != nil {
+		if request.Device.IP != "" {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		} else if request.Device.IPv6 != "" {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+	}
+	if request.Site != nil {
+		headers.Add("Referer", request.Site.Page)
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     url,
+		Headers: headers,
+		Body:    body,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	var httpRequests []*adapters.RequestData
+	var tempErrors []error
+	gdprApplies, consentString := readGDPR(request)
+
+	var missenaInternalParams MissenaInternalParams
+
+	for _, imp := range request.Imp {
```

Comment:
It appears you are prepared to handle multiple imps but really you only care that at least one of them is valid as you're just preparing a single request without specifying the impression ID, though you are grabbing the params from the last imp you process. Is this what you want or do you want to grab the info from the first imp?

I suggest adding a test where there are multiple imps with different params, specifically `placement`. I also think you should add a test where one imp results in an error and the other is well formed.

Reply:
Done ✅